base10_test.go

```diff
@@ -0,0 +1,266 @@
+package uint256
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+	"testing"
+)
+
+func TestStringScanBase10(t *testing.T) {
+	z := new(Int)
+
+	type testCase struct {
+		i   string
+		err error
+		val string
+	}
+
+	cases := []testCase{
+		{i: twoPow256Sub1 + "1", err: ErrBig256Range},
+		{i: "2" + twoPow256Sub1[1:], err: ErrBig256Range},
+		{i: twoPow256Sub1[1:]},
+		{i: twoPow128},
+		{i: twoPow128 + "1"},
+		{i: twoPow128[1:]},
+		{i: twoPow64 + "1"},
+		{i: twoPow64[1:]},
+		{i: "banana", err: ErrSyntaxBase10},
+		{i: "0xab", err: ErrSyntaxBase10},
+		{i: "ab", err: ErrSyntaxBase10},
+		{i: "0"},
+		{i: "000", val: "0"},
+		{i: "010", val: "10"},
+		{i: "01", val: "1"},
+		{i: "-0", err: ErrSyntaxBase10},
+		{i: "-10", err: ErrSyntaxBase10},
+		{i: "115792089237316195423570985008687907853269984665640564039457584007913129639936", err: ErrBig256Range},
+		{i: "115792089237316195423570985008687907853269984665640564039457584007913129639935"},
+	}
+
+	for _, v := range cases {
+		err := z.SetFromBase10(v.i)
+		if !errors.Is(err, v.err) {
+			t.Errorf("expect err %s, got %s", v.err, err)
+		}
+		if err == nil {
+			got := z.ToBig().String()
+			want := v.i
+			if v.val != "" {
+				want = v.val
+			}
+			if got != want {
+				t.Errorf("expect val %s, got %s", v.i, got)
+			}
+		}
+	}
+}
+
+func FuzzBase10StringCompare(f *testing.F) {
+	var (
+		bi        = new(big.Int)
+		z         = new(Int)
+		max256, _ = FromBase10(twoPow256Sub1)
+	)
+	for _, tc := range []string{
+		twoPow256Sub1 + "1",
+		"2" + twoPow256Sub1[1:],
+		twoPow256Sub1,
+		twoPow128,
+		twoPow128,
+		twoPow128,
+		twoPow64,
+		twoPow64,
+		"banana",
+		"0xab",
+		"ab",
+		"0",
+		"000",
+		"010",
+		"01",
+		"-0",
+		"-10",
+		"115792089237316195423570985008687907853269984665640564039457584007913129639936",
+		"115792089237316195423570985008687907853269984665640564039457584007913129639935",
+		"apple",
+		"04112401274120741204712xxxxxz00",
+		"0x10101011010",
+		"熊熊熊熊熊熊熊熊",
+	} {
+		f.Add(tc)
+	}
+	f.Fuzz(func(t *testing.T, orig string) {
+		err := z.SetFromBase10(orig)
+		val, ok := bi.SetString(orig, 10)
+		// if fail, make sure that we failed too
+		if !ok {
+			if err == nil {
+				t.Errorf("expected base 10 parse to fail: %s", orig)
+			}
+			return
+		}
+		// if its negative number, we should err
+		if len(orig) > 0 && (orig[0] == '-') {
+			if !errors.Is(err, ErrSyntaxBase10) {
+				t.Errorf("should have errored at negative number: %s", orig)
+			}
+			return
+		}
+		// if its too large, ignore it also
+		if val.Cmp(max256.ToBig()) > 0 {
+			return
```

Comment:
If >256, we should have a non-nil `err`. Should check that

Reply:
added this